[WIP] feat: (Core) Introduce user actions menu

mykolavorobiov opened the pull request:
#### Please provide a link to the associated issue.

closes https://github.com/SAP/fundamental-ngx/issues/4301
fixes https://github.com/SAP/fundamental-ngx/issues/5048

#### Please provide a brief summary of this pull request.

User Actions Menu - develop the component in ngx-core
https://experience.sap.com/fiori-design-web/launchpad/#user-menu

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] Documentation Examples
- [x] Stackblitz works for all examples

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.scss:
```diff
@@ -0,0 +1,31 @@
+            left: 0;
+            width: 100%;
+            height: 100%;
+            object-fit: scale-down;
```
`object-fit` is not supported by IE11, isn't?

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+  templateUrl: './user-actions-menu-item.component.html',
+  styleUrls: ['./user-actions-menu-item.component.scss'],
+  host: {
+    '[class.fd-user-actions-menu-item]': 'true',
```
we could use just `'class': 'fd-user-actions-menu-item'` to avoid binding.

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+  glyphSrc: string;
+
+  @Input()
+  routerLink: string | any[];
```
Is there a reason to have this input here cause `routerLink` works by itself?

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+    if (this._activeDialog) {
+      this._dismissDialog();
+    }
+    this.onOpenDialog.emit(this._activeDialog);
```
At this point `_activeDialog` doesn't exist yet;

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+    this.onOpenDialog.emit(this._activeDialog);
+    this._activeDialog = this._dialogService.open(this.dialogContent, { responsivePadding: true });
+    this._activeDialog.afterClosed.pipe(
+      take(1)
```
`afterClosed` fires only once by itself, no need to use `take`

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+  /** @hidden */
+  private _dismissDialog(): void {
+    if (this._activeDialog) {
+      this._activeDialog.dismiss();
```
`dialogRef.dismiss()` will send an error to `afterClosed` stream and subscription on line `79` wont be called.
So if it's needed to reset `this._activeDialog` we need to listen for error as well.

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu/user-actions-menu.component.spec.ts:
```diff
@@ -0,0 +1,26 @@
+
+import { UserActionsMenuComponent } from './user-actions-menu.component';
+
+fdescribe('UserActionsMenuComponent', () => {
```
please remove focus mode

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu/user-actions-menu.component.ts:
```diff
@@ -0,0 +1,105 @@
+import { ChangeDetectionStrategy, Component, ContentChildren, EventEmitter, Input, Output, QueryList, ViewChild, ViewEncapsulation } from '@angular/core';
+import { Placement, PopperOptions } from 'popper.js';
```
As far as I know we a moving away from popper.js https://github.com/SAP/fundamental-ngx/issues/2527
@JKMarkowski could you help with it?

JKMarkowski replied:
Feel free to just import it from `libs/core/src/lib/popover/popover-position/popover-position.ts`, instead of having popper.js ref

dimamarksman commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu/user-actions-menu.component.ts:
```diff
@@ -0,0 +1,105 @@
+  @ContentChildren(UserActionsMenuItemComponent) items: QueryList<UserActionsMenuItemComponent>;
+
+  /** menu open state */
+  isOpenMenu = false;
+  hasItems = false;
```
Does anyone use these properties?

dimamarksman reviewed (changes requested):
LGTM
Added several comments, please take a look

JKMarkowski commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+  styleUrls: ['./user-actions-menu-item.component.scss'],
+  host: {
+    '[class.fd-user-actions-menu-item]': 'true',
+    '[attr.tabindex]': 'tabIndex'
```
There is no `tabIndex` in this file

JKMarkowski commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class UserActionsMenuItemComponent implements OnDestroy {
+  @Input()
```
Could you add some descriptions into inputs?

JKMarkowski commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+  /** @hidden Active dialog */
+  private _activeDialog: DialogRef;
+
+  @HostListener('click')
```
Could you hide it from docs/

JKMarkowski commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+    this._activeDialog = this._dialogService.open(this.dialogContent, { responsivePadding: true });
+    this._activeDialog.afterClosed.pipe(
+      take(1)
+    ).subscribe(() => {
```
Let's remove unnecessary brackets

JKMarkowski commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.ts:
```diff
@@ -0,0 +1,90 @@
+  routerLink: string | any[];
+
+  @Input()
+  dialogContent: DialogContentType;
```
I see it works well, but I'm wondering if there is any sense of having this here, instead of just opening dialog by application side.

mykolavorobiov replied:
I added it only for simple content, for example, `About` or something like this. Without any business logic or functionality into them. But of course, you as a developer can add click handler and add any functionality to this item.

JKMarkowski commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu/user-actions-menu.component.ts:
```diff
@@ -0,0 +1,105 @@
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class UserActionsMenuComponent {
```
If you want to get rid of all of these inputs, you can extend base popover class from `libs/core/src/lib/popover/base/base-popover.class.ts`.

JKMarkowski reviewed (dismissed):
Hello @mykolavorobiov Examples work well, good job!  I added some comments there. I'm wondering about changing all of `public_api` imports to direct imports, just to avoid weird errors and circular dependencies. let me know what you think about it

Seamoo13 commented:
- [ ] 1. please add tooltip for truncated items
![2021-01-29_3](https://user-images.githubusercontent.com/47522152/106266142-bf28b380-6230-11eb-8c05-7a998bb43ca1.png)

- [ ] 2. shouldn't sign out be the last option? or is it up to the developer/user to decide?
![2021-01-29_4](https://user-images.githubusercontent.com/47522152/106272579-eb94fd80-6239-11eb-9dc4-23955f17b6a0.png)

mykolavorobiov commented:
@Seamoo13 
1) Ok, will do it
2) It's an example.

dimamarksman commented:
@mykolavorobiov thank you for addressing my comments.
There are still two without answer, could you please share your thoughts regarding it.

EvilAlexei commented on apps/docs/src/app/core/component-docs/shellbar/examples/shellbar-collapsible-example.component.html:
```diff
@@ -39,6 +39,63 @@
+
+            <fd-user-actions-menu [user]="user">
+                <fd-user-actions-menu-item
+                    [title]="'Looonnngg title Looonnngg title Looonnngg title Looonnngg title Looonnngg title Looonnngg title Looonnngg title'"
```
Nit: Maybe it will be better to use something more readable? Default `Lorem ipsum...` :)?

EvilAlexei commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu-item/user-actions-menu-item.component.scss:
```diff
@@ -0,0 +1,19 @@
+.fd-user-actions-menu-item {
+    &__custom-icon.fd-list__icon[class*=sap-icon] {
```
I think `^` will be better and there will be less chance for name collision.
```suggestion
    &__custom-icon.fd-list__icon[class^=sap-icon] {
```

EvilAlexei commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu/user-actions-menu.component.scss:
```diff
@@ -0,0 +1,12 @@
+.fd-user-actions-menu {
+    &__body {
+        min-width: 10rem;
+        max-width: 20rem;
+        .fd-user-actions-menu__user-name {
```
Please add empty space between rules and selectors :)

EvilAlexei commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu/user-actions-menu.component.scss:
```diff
@@ -0,0 +1,12 @@
+        max-width: 20rem;
+        .fd-user-actions-menu__user-name {
+            padding: 1rem 1rem 0.5rem 1.25rem;
+        }
+        .fd-list {
```
Please add empty space between brace and selector :)

mikerodonnell89 reviewed (commented):
What specs are you basing this off? Looks quite a bit different from what I'm seeing on the wiki.  I believe we will likely need some new classes in fundamental-styles before we can work on implementing this in NGX

Some issues in the screenshot below:

- [ ] name should be centered (I know this is not how it is currently on Fiori, but according to the specs it should be) 

- [ ] empty icon next to custom SAP glyph

- [ ] custom glyph should be scaled and have some space between the text

- [ ] popover alignment incorrect (should not be centered)

![Screen Shot 2021-02-02 at 2 17 31 PM](https://user-images.githubusercontent.com/2471874/106664319-219bef80-6562-11eb-8365-38a7a6f95755.png)

Betrozov commented on apps/docs/src/app/core/component-docs/shellbar/shellbar-docs.component.ts:
```diff
@@ -63,4 +69,18 @@ export class ShellbarDocsComponent {
+        {
+            language: 'html',
+            code: shellbarUserActionsMenuHTMLSrc,
+            fileName: 'shellbar-user-actions-menu-example.component'
```
`shellbar-user-actions-menu-example.component` -> `shellbar-user-actions-menu-example`

Betrozov commented on apps/docs/src/app/core/component-docs/shellbar/shellbar-docs.component.ts:
```diff
@@ -63,4 +69,18 @@ export class ShellbarDocsComponent {
+            language: 'typescript',
+            component: 'ShellbarUserActionsMenuExample',
+            code: shellbarUserActionsMenuTsSrc,
+            fileName: 'shellbar-user-actions-menu-example.component'
```
the same

Betrozov commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu.module.ts:
```diff
@@ -0,0 +1,29 @@
+    ButtonModule
+  ]
+})
+export class UserActionsMenuModule { }
```
Can you please add `UserActionsMenuModule` to the `FundamentalNgxCoreModule`?

mykolavorobiov replied:
If we add to `FundamentalNgxCoreModule`, we should create separate doc and examples for this. But for now it's only a part of shellbar module.

Betrozov replied:
oh, ok, thanks for the explanation

mykolavorobiov commented:
@mikerodonnell89 Regarding to your comment
- https://app.gitbook.com/@fundamental-ngx/s/docs/ready-for-review/user-actions-menu-core this is tech doc for this component.
- add some spaces (0.5rem) between icon and text 
- fixed empty icon (wrong suggestion from PR reviewer)
- changed position for menu. You can provide your position, same as popover (this menu extends `BasePopover`. It's a non-highlighted option in doc).

InnaAtanasova commented:
@mykolavorobiov this is totally not your fault but that's not how the component should look like. I am not allowed to share links here to the specs, but please check in Wiki `Central Home/User Menu`. I will send you in PM the link

mikerodonnell89 commented:
Couple small things:
also @InnaAtanasova should the styles for this come from the fundamental-styles repo?

- [ ] Strange focus outline after clicking user menu:
![Screen Shot 2021-03-31 at 12 28 26 PM](https://user-images.githubusercontent.com/2471874/113192918-9a46c180-921c-11eb-85d2-a30190152f38.png)

- [ ] No margin/padding between the back button in the submenu and the SubmenuLevel 1 text
![Screen Shot 2021-03-31 at 12 35 26 PM](https://user-images.githubusercontent.com/2471874/113193769-949dab80-921d-11eb-9491-3e74a503962e.png)

- [ ] Compact user menu - on first page, sign out and the sap-icon--grid button in the top right are not compact (just a documentation thing)

mikerodonnell89 reviewed (changes requested):
some small things, see previous comment

mikerodonnell89 commented:
Still seeing something weird here

![Screen Shot 2021-04-02 at 11 59 42 AM](https://user-images.githubusercontent.com/2471874/113441120-ec1d5200-93aa-11eb-8e6b-f940269b8719.png)

![Screen Shot 2021-04-02 at 12 00 16 PM](https://user-images.githubusercontent.com/2471874/113441149-00614f00-93ab-11eb-8a62-1f96a9b8a619.png)

Also it might be best to leave the old shellbar-user-menu component for now, with a note of it being deprecated, so as to not introduce breaking changes without warning, and we can open a ticket to remove it in some later version

mikerodonnell89 commented:
Did not mean to click close

Betrozov commented:
> Still seeing something weird here
> 
> ![Screen Shot 2021-04-02 at 11 59 42 AM](https://user-images.githubusercontent.com/2471874/113441120-ec1d5200-93aa-11eb-8e6b-f940269b8719.png)
> 
> ![Screen Shot 2021-04-02 at 12 00 16 PM](https://user-images.githubusercontent.com/2471874/113441149-00614f00-93ab-11eb-8a62-1f96a9b8a619.png)

Hi @mikerodonnell89 
So in the examples, I am using `fd-list` to define the menu list items and using a popup with `[focusAutoCapture]="true"`.
this means that `focusAutoCapture` is looking for the first element that can be focused and that is `fd-list`, namely the `ul` element
The `ul` element has default focus styles

droshev commented:
this is a wrong example with 2 user menus
![Screen Shot 2021-04-05 at 3 27 33 PM](https://user-images.githubusercontent.com/4380815/113616733-7bcd3580-9623-11eb-917b-38c5ac3df19d.png)

droshev reviewed (changes requested):
@mykolavorobiov this PR looks like having a breaking change. Is this intentional? If so then we need to document that

InnaAtanasova reviewed (changes requested):
User Menu is part of ShellBar but a separate component and its css belongs to Fundamental-styles. Please open a PR there where we can properly test the css and then we can review the PR in fund-ngx.

Betrozov commented:
> User Menu is part of ShellBar but a separate component and its css belongs to Fundamental-styles. Please open a PR there where we can properly test the css and then we can review the PR in fund-ngx.

@InnaAtanasova will be done

Betrozov commented:
> @mykolavorobiov this PR looks like having a breaking change. Is this intentional? If so then we need to document that

Hi @droshev 
Will revert 

> @mykolavorobiov this PR looks like having a breaking change. Is this intentional? If so then we need to document that

Hi @droshev 
going to revert the `shellbar-user-menu` component for now, with a note of it being deprecated, so as to not introduce breaking changes without warning and going to open an issue to remove it in some later version

InnaAtanasova commented:
Added [WIP] and `blocked` label as we are waiting for styles PR to be merged first.

kavya-b commented on apps/docs/src/app/core/component-docs/shellbar/examples/shellbar-user-actions-menu-compact-example.component.html:
```diff
@@ -0,0 +1,106 @@
+        <li fd-list-item fd-user-actions-menu-item>
+            <a fd-list-link>
+                <i fd-list-icon glyph="border"></i>
+                <span fd-list-title> Lorem Ipsum is simply dummy text of the printing and typesetting industry </span>
```
hovering over this long text should show the entire text in a tooltip as it appears truncated.

kavya-b commented on apps/docs/src/app/core/component-docs/shellbar/examples/shellbar-user-actions-menu-example.component.html:
```diff
@@ -0,0 +1,58 @@
+                    <a fd-list-link>
+                        <i fd-list-icon glyph="border"></i>
+                        <span fd-list-title>
+                            Lorem Ipsum is simply dummy text of the printing and typesetting industry
```
same as above, `title` may be needed to show the entire text in tooltip because it is truncated.

kavya-b commented on apps/docs/src/app/core/component-docs/shellbar/examples/shellbar-user-actions-menu-header-example.component.ts:
```diff
@@ -0,0 +1,35 @@
+        fullName: 'John Doe',
+        image: 'https://i.pravatar.cc/150?img=2',
+        colorAccent: 1,
+        role: 'User Experience Designe'
```
```suggestion
        role: 'User Experience Designer'
```

kavya-b commented on libs/core/src/lib/fundamental-ngx-core.module.ts:
```diff
@@ -89,6 +89,7 @@ import { DynamicPageModule } from './dynamic-page/dynamic-page.module';
 import { StatusIndicatorModule } from './status-indicator/status-indicator.module';
 import { GridListModule } from './grid-list/grid-list.module';
 import { FacetModule } from './facets/facet.module';
+import { UserActionsMenuModule } from './shellbar/public_api';
```
can this refer to the [.module instead of public_api](https://github.com/SAP/fundamental-ngx/pull/4502#pullrequestreview-578270438)?

kavya-b commented on libs/core/src/lib/shellbar/user-actions-menu/components/user-actions-menu-footer/user-actions-menu-footer.component.ts:
```diff
@@ -0,0 +1,55 @@
+        /** @hidden */
+        private readonly _cd: ChangeDetectorRef,
+        /** @hidden */
+        @Optional() @Host() private readonly _manuComponent: UserActionsMenuComponent,
```
```suggestion
        @Optional() @Host() private readonly _menuComponent: UserActionsMenuComponent,
```

kavya-b commented on libs/core/src/lib/shellbar/user-actions-menu/components/user-actions-menu-footer/user-actions-menu-footer.component.ts:
```diff
@@ -0,0 +1,55 @@
+        /** @hidden */
+        @Optional() @Host() private readonly _manuComponent: UserActionsMenuComponent,
+        /** @hidden */
+        @Optional() @Host() private readonly _submanuComponent: UserActionsSubmenuComponent
```
```suggestion
        @Optional() @Host() private readonly _submenuComponent: UserActionsSubmenuComponent
```

kavya-b commented on libs/core/src/lib/shellbar/user-actions-menu/components/user-actions-menu-header/user-actions-menu-header.component.ts:
```diff
@@ -0,0 +1,94 @@
+        /** @hidden */
+        private readonly _cd: ChangeDetectorRef,
+        /** @hidden */
+        @Optional() @Host() private readonly _manuComponent: UserActionsMenuComponent,
```
```suggestion
        @Optional() @Host() private readonly _menuComponent: UserActionsMenuComponent,
```
same with `_submanuComponent` --> `_submenuComponent`

kavya-b commented on libs/core/src/lib/shellbar/user-actions-menu/components/user-actions-submenu/user-actions-submenu.component.ts:
```diff
@@ -0,0 +1,72 @@
+        );
+    }
+
+    ngOnDestroy(): void {
```
`@hidden`?

kavya-b commented on libs/core/src/lib/shellbar/user-actions-menu/directives/user-actions-menu-item.directive.ts:
```diff
@@ -0,0 +1,91 @@
+    selector: '[fd-user-actions-menu-item]'
+})
+export class UserActionsMenuItemDirective implements OnDestroy {
+    @Input()
```
some descriptions for the inputs would be nice.

kavya-b commented on libs/core/src/lib/shellbar/user-actions-menu/directives/user-actions-menu-item.directive.ts:
```diff
@@ -0,0 +1,91 @@
+} from '@angular/core';
+import { finalize } from 'rxjs/operators';
+
+import { DialogContentType, DialogRef, DialogService } from '../../../dialog/public_api';
```
can we import this directly,[ instead of public_api](https://github.com/SAP/fundamental-ngx/pull/4502#pullrequestreview-578270438)?

Betrozov replied:
@kavya-b Why is it bad to import from public_api?

kavya-b replied:
there was an issue sometime back where in ng10, using public_api for imports was causing some errors while running `ng xi18n`. Here is the PR where that issue was fixed: [link](https://github.com/SAP/fundamental-ngx/pull/4889)

kavya-b commented on libs/core/src/lib/shellbar/user-actions-menu/user-actions-menu.module.ts:
```diff
@@ -0,0 +1,45 @@
+        DialogModule,
+        ButtonModule,
+        BarModule
+    ]
```
do we need to provide the UserActionsMenuService also here?

Betrozov replied:
no, it is already provided [here](https://github.com/SAP/fundamental-ngx/blob/fb00fffd81f6d36e4636fc3dceb1daaba18bcb8c/libs/core/src/lib/shellbar/user-actions-menu/components/user-actions-menu/user-actions-menu.component.ts)

kavya-b reviewed (changes requested):
Nice work, few comments below. Please take a look!
- on opening the menu, a blue border focus immediately appears on the actions; is this expected?
![image](https://user-images.githubusercontent.com/53509521/116335499-9ba3e380-a7f4-11eb-8365-8c290a4f2356.png)

InnaAtanasova commented:
<img width="357" alt="Screen Shot 2021-04-28 at 2 38 22 PM" src="https://user-images.githubusercontent.com/39598672/116455813-826c5880-a82f-11eb-97ee-9ca92201b0bb.png">

The issues found by Kavya are still there. The menu can't be navigated by tab.

InnaAtanasova commented:
<img width="785" alt="Screen Shot 2021-04-28 at 2 40 55 PM" src="https://user-images.githubusercontent.com/39598672/116456019-c19aa980-a82f-11eb-8427-ad35057a28f7.png">
The extra description is not truncated.

<img width="309" alt="Screen Shot 2021-04-28 at 2 45 23 PM" src="https://user-images.githubusercontent.com/39598672/116456524-61583780-a830-11eb-819a-562aa8f5c67e.png">
RTL is not applied for the button in the header.

InnaAtanasova commented:
@Betrozov the focus issue is still there. When you start navigating with the Tab key the focus is lost after the second item in the list (Router Link).

kari-leg commented on apps/docs/src/app/core/component-docs/shellbar/shellbar-docs.component.html:
```diff
@@ -45,20 +48,101 @@
-    How it works
+<fd-docs-section-title id="user-actions-menu" componentName="shellbar"> User actions menu </fd-docs-section-title>
+<description>
+    This example shows the user actions menu in shellbar. If no user thumbnail is available then display initials.
```
rephrase to: 
Clicking on the profile picture opens the user actions menu. If no profile picture is set, a default icon is displayed with initials. Once clicked on the icon, options are vertically listed: [Recent Activities, Frequently Used, App Finder, Settings, Contact Support, Give Feedback, About, App Settings and Sign out]
<img width="279" alt="Screen Shot 2021-05-20 at 5 10 54 PM" src="https://user-images.githubusercontent.com/84039792/119049366-6e2efd80-b98e-11eb-9035-f6b2ca9223cd.png">
.

kari-leg commented on apps/docs/src/app/core/component-docs/shellbar/shellbar-docs.component.html:
```diff
@@ -45,20 +48,101 @@
+<separator></separator>
+
+<fd-docs-section-title id="user-actions-menu-compact" componentName="shellbar">
+    User actions menu compact mode
```
When clicked on ‘example click,’ ‘SubMenu level 2’ is defined and further on ‘SubMenu level 3’ appears with the sign out option being visible at the footer.
